stardis/plasma/base.py

```diff
@@ -129,6 +129,149 @@ def calculate(
+        Alpha calculation for each line from Vald at each depth point.
+        See Rybicki and Lightman eq. 1.80. Voigt profiles are calculated later, and
+        B_12 is substituted appropriately out for f_lu.
+        Assumes LTE for lower level population.
+    """
+
+    outputs = ("alpha_line_from_linelist", "lines_from_linelist")
+    latex_name = (r"\alpha_{\textrm{line, vald}}",)
+    latex_formula = (
+        r"\dfrac{\pi e^{2} n_{lower} f_{lu}}{m_{e} c}\
+        \Big(1-exp(-h \nu / k T) \phi(\nu)\Big)",
+    )
+
+    def calculate(
+        self,
+        atomic_data,
+        ion_number_density,
+        t_electrons,
+        g,
+        ionization_data,
+    ):
+        # solve n_lower : n * g_i / g_0 * e ^ (-E_i/kT)
+        # get f_lu : loggf -> use g = 2j+1
+        # emission_correction = (1-e^(-h*nu / kT))
+        # alphas = ALPHA_COEFFICIENT * n_lower * f_lu * emission_correction
+
+        points = len(t_electrons)
+
+        # Need degeneracy of ground state of the ion to calculate n_lower
+        linelist = atomic_data.linelist.rename(columns={"ion_charge": "ion_number"})[
+            [
+                "atomic_number",
+                "ion_number",
+                "wavelength",
+                "log_gf",
+                "e_low",
+                "e_up",
+                "j_lo",
+                "j_up",
+                "rad",
+                "stark",
+                "waals",
+            ]
+        ].merge(
+            g.loc[:, :, 0].rename("g_0"),
+            how="left",
+            on=["atomic_number", "ion_number"],
+        )
+
+        # Truncate to final atomic number
+        linelist = linelist[
+            linelist.atomic_number <= (atomic_data.selected_atomic_numbers.max())
+        ]
+
+        # Calculate degeneracies
+        linelist["g_lo"] = linelist.j_lo * 2 + 1
+        linelist["g_up"] = linelist.j_up * 2 + 1
+
+        exponent_by_point = np.exp(
+            np.outer(
+                -linelist.e_low.values * u.eV, 1 / (t_electrons * u.K * const.k_B)
+            ).to(1)
+        )
+
+        # grab densities for n_lower - need to use linelist as the index
+        linelist_with_densities = linelist.merge(
+            ion_number_density,
+            how="left",
+            on=["atomic_number", "ion_number"],
+        )
+
+        n_lower = (
+            (
+                exponent_by_point * linelist_with_densities[np.arange(points)]
+            ).values.T  # arange mask of the dataframe returns the set of densities of the appropriate ion for the line at each point
+            * linelist_with_densities.g_lo.values
+            / linelist_with_densities.g_0.values
+        )
+
+        linelist["f_lu"] = (
+            10**linelist.log_gf / linelist.g_lo
+        )  # vald log gf is "oscillator strength f times the statistical weight g of the parent level"  see 1995A&AS..112..525P
+
+        line_nus = (linelist.wavelength.values * u.AA).to(
```

Comment:
join with 249

Reply:
I realized they can't be joined with 249 because you need the line_nus for the emission correction calculation. Since they go in there I think it makes more sense to calculate them once and attach them where appropriate, rather than do the conversion from wavelengths to frequency 3 separate times.